### PR TITLE
Move code analyzers to arcade and add analyzer features to reference them in Tools.props and filter analyzers

### PR DESCRIPTION
**Pull request by safern**
This ports the analyzers that were part of buildtools and also enables repos to add package reference to packages containing analyzers whenever does repos don't restore per project (coreclr, corefx, wcf, etc).

Also, this enables repos that consume the SDK to filter analyzers in the sense of repos that restore once and not per project, but they don't use the same set of analyzers for all the projects. So for example in corefx we're going to add Xunit analyzers, so this enables us to add that package reference in Tools.props directly and then in our props file if the project built is a test project we can set the item `AnalyzersToInclude` to contain this analyzer and therefore only that one will be passed to the compiler. i.e:

```xml
<ItemGroup>
  <AnalyzersToInclude Include="Microsoft.DotNet.CodeAnalysis" />
</ItemGroup>
```

So for that project it will only pass in `Microsoft.DotNet.CodeAnalysis` analyzer to the compiler.

### CODE REVIEW

**Comment by ericstj** [comment]
> why are these analyzers part of infrastructure?

They have their own package I think.  They're all about enforcing conventions we follow in CoreFx, WCF, and CoreCLR.  Probably some can be cleaned up, perhaps some can be shared more generally, but I thought that was what arcade was about?  We can't just put all analyzers in a single place, just like we don't put all libraries in CoreFx.

> Tools.props is not the right place to reference packages that are used during inner loop

This is the pattern that @weshaggard already established for restore-once repositories like CoreFx.  We need to have a pattern for restore-once and we shouldn't rehash that pattern in every PR that extends it.

Are you disagreeing with the premise of restore-once or are you disagreeing with the implementation of using `Tools.props` for that restore?  We need to get on the same page here as this will be a recurring pattern when we bring things over from buildtools.

**Comment by markwilkie** [comment]
> I thought that was what arcade was about?

Yes.  So far we've agreed that the requirements are less stringent for the so-called "optional" packages (those which are not common or required for all).  Of course, reasonable effort should be made to take the opportunity to clean up, and make reuse reasonable for other repos, and the code should not be a "hack".....  But the optional packages *do not* have to necessarily work for everyone - but again, every reasonable effort should be made to get at close to that ideal as possible.

Also, I have work to do to make more clear and re-enforce that the existence of a package/tool in arcade does *not* indicate ownership.  So far what's clear is that the "required" bits of Arcade (I need to also update the docs w/ the latest) is owned by eng svcs, with the awesome help and guidance from the arcade working group.

**Comment by tmat** [comment]
@ericstj I disagree specifically with adding analyzers to `Tools.proj`. Right now I don't see why they can't be specified as regular `PackageReference`s in the projects that use them.

More generally, I don't like "restore-once" approach. I understand some core repos might require it, but would rather keep that set of repos as small as possible, ideally just CoreFX. The reasoning why CoreFX needs it seems sound to me - it builds the core platform packages. This reason might apply to CoreCLR as well, but I don't know if it actually does. I'd definitely question that WCF requires it - I don't see why it is special in any way. Why is it different from Roslyn, symreader, dotnet SDK, etc. repos that do not need "restore-only".

BTW, Arcade SDK already has extensibility point that allows repos to add more package references to "Tools" restore phase without requiring changes in the SDK. You can add `Tools.props` to the `eng` directory of the repo and it gets imported to `Tools.proj`. See

https://github.com/dotnet/arcade/blob/master/Documentation/ArcadeSdk.md#engtoolsprops-optional

So I'd prefer repos that need/want to restore some more packages to use that mechanism without updating the SDK.

**Comment by safern** [comment]
So here I'm adding one package containing custom analyzers used in corefx, coreclr, wcf and other repos. The package is Microsoft.DotNet.CodeAnalysis -- which contains a set of analyzers we use, there is a couple that I think could be deleted like `MembersMustExistAnalyzer`. There is definitely cleanup that can be made in the Default.ruleset that is included as well, I will invest some time on cleaning that up.

The other piece that I'm adding are 2 targets to the SDK (that are disabled by default and repos can enable it through a property) -- and I think they will be of common usage in at least 3 or 4 repos that we already mention follow the same pattern on analyzers due to their restore process. The second target enables this kind of projects to filter analyzers by setting an itemgroup per project or project kind.

> BTW, Arcade SDK already has extensibility point that allows repos to add more package references to "Tools" restore phase without requiring changes in the SDK. You can add Tools.props to the eng directory of the repo and it gets imported to Tools.proj. See

Yes that is the pattern that we're following in corefx, however to follow that pattern in corefx we needed this targets for analyzers to be added. We thought they would be common enough to try to push them into the SDK and that is why we disable them by default. Then repos can enable this feature through a property, which I thought it was OK, to have common tools that can be opted-in as part of the SDK. We can definitely maybe put this pattern directly in another package and have CoreFX and other repos that might need it consume it, but then we would need somewhere to put the `SaveItem` task there.

About the SaveItem task @ericstj and I agree that it is a very general task that can be used by multiple repos, and we even think it is so general that it could even be part of msbuild itself -- so that is why we think it is a nice to have task as part of the arcade SDK.

**Comment by tmat** [comment]
@safern I'm still not hearing why the analyzer package can't be referenced directly by the projects, which is the standard approach.

**Comment by safern** [comment]
> @safern I'm still not hearing why the analyzer package can't be referenced directly by the projects, which is the standard approach.

Because in CoreFX we don't restore by project. In order for the analyzers to be resolved by nuget the referenced analyzer package needs to be restored so that the project has an assets file and then nuget will read it from that assets file from the package contents section and populate the analyzer item which is then passed to the compiler, right? 

If we want to do that, we would need to change CoreFX to restore about 700 hundred projects that we have, I think that would have a significant impact to our build time.

**Comment by safern** [comment]
Btw, we're still using the build system to resolve the analyzers, the would get added as a package reference in Tools.props and then resolve them through the nuget task to resolve package assets and we would capture the Analyzer item in a static props file. So still using that infrastructure, just capturing early in order to avoid restoring all the projects in order to get that item populated with the required analyzers.

**Comment by tmat** [comment]
@safern Ok, so now I see why you want this for CoreFX and seems reasonable as a workaround for a perf issue. We should definitely follow up on why restoring 700 projects is slow. Customers have solutions with hundreds of projects, so if it doesn't work for us it won't work for them either. So this needs to be addressed. It would be good to have some numbers and an issue tracking this perf problem.

Now why is this an issue in other repos? Do we have 700 projects in WCF repo as well?

**Comment by ericstj** [comment]
> Now why is this an issue in other repos? Do we have 700 projects in WCF repo as well?

I've actually been trying to convince WCF to get off of this build system ever since the start of 2.0.  They needed something in 1.0 because we didn't yet have a public toolchain, so they copied CoreFx.  They've stayed on it because they got too invested in the complexity to untangle themselves: hard to justify the larger change to switch when they can keep paying small change to stay on.  Maybe that tradeoff will tip the other direction with the Arcade conversion.

> We should definitely follow up on why restoring 700 projects is slow

I spent a good amount of time looking into this in 1.0.  Part of it was due to https://github.com/NuGet/Home/issues/3087 which we worked around, and was later fixed by NuGet.  Another issue in 1.0 was needing to spin up NuGet process multiple times, which should be fixed now that restore runs in-proc in MSBuild.  Another problem (not incremental, but clean restore) was due to the overall cost to write out all the assets files: ultimately this was what forced our hand in 2.0.  Nuget doesn't have anything to cache redundant work across projects, even when the work is completely repetitive.

**Comment by tmat** [comment]
@ericstj Thanks for writing up the history. Sounds like there is still work in NuGet to be done to improve the perf. Do you know if there is a work item tracking it?

Given this information I'd suggest keeping this mechanism of adding analyzers in CoreFX repo, instead of adding it to Arcade SDK. I believe Arcade SDK has enough extensibility points to make it possible (the `Tools.prop` file I mentioned above).

Re WCF - we should definitely push on removing the complexity. I think now is the right time. Let's file an issue to do that. If we want something that works short term, we can duplicate the targets that populate the analyzer items in CoreFX and WCF repos. It's not that much code. 

I'm fine having `SaveItems` task in the SDK.

**Comment by safern** [comment]
> I believe Arcade SDK has enough extensibility points to make it possible (the Tools.prop file I mentioned above).

I agree with this. However, would you mind importing the BuildTask.props in Tools.proj? So that I can have access to the `ArcadeSdkBuildTasksAssembly` and not have to find the path to the SDK ourselves because we're going to use the `SaveItem` task directly to generate the analyzer's props file. 

> Re WCF - we should definitely push on removing the complexity. I think now is the right time. Let's file an issue to do that. If we want something that works short term, we can duplicate the targets that populate the analyzer items in CoreFX and WCF repos. It's not that much code.

I think that could just be mentioned in: https://github.com/dotnet/arcade/issues/919

I'm fine with not putting these targets in the SDK as it is a small piece of code that can be replicated if needed for now. So at the end I will update my PR to contain the following changes:

- Add `SaveItem` task into the SDK.
- Expose the `ArcadeSdkBuildTasksAssembly` property in `Tools.proj` to have access to that task. 
- add `Microsoft.DotNet.CodeAnalysis package for our custom analyzers.

Sounds fair?

**Comment by tmat** [comment]
> I agree with this. However, would you mind importing the BuildTask.props in Tools.proj? 

Yeah, no problem with that.

File: src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+using Microsoft.Build.Framework;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Utilities;
+using System.Collections.Generic;
+using System.Linq;
+using System.IO;
+
+namespace Microsoft.DotNet.Arcade.Sdk
+{
+    public class SaveItems : Task
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs]: >SaveItems [](start = 17, length = 9)

Please, add some documentation what this task is good for.

File: src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        [Required]
+        public string ItemName { get; set; }
+
+        [Required]
+        public ITaskItem[] Items { get; set; }
+
+        [Output]
+        [Required]
+        public string[] Files { get; set; }
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs]: >Files [](start = 24, length = 5)

Should this just take a single file? I don't see why I would want to write the same content to multiple files.

**Comment by safern** [comment]
> You mean add the sources of this package to Arcade repo? If so then yes, that's fine as well. I was wondering if some of those rules were more general purpose and applicable outside of CoreFX repo we could port them to FxCop analyzers. If not then having them in Arcade repo is ok.

Yes that is what I meant. I think this is more CoreFX/Clr/WCF like analyzers, but they could probably be tweaked to make them more general purpose, in that case we could port them and delete the package, no?

File: src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs (hunk 1 above)
Comment (safern) [reply src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs]: Sure.

File: src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs (hunk 2 above)
Comment (safern) [reply src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs]: Yes, that sounds a reasonable change.

File: src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+using Microsoft.Build.Framework;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Utilities;
+using System.Collections.Generic;
+using System.Linq;
+using System.IO;
+
+namespace Microsoft.DotNet.Arcade.Sdk
+{
+    // This task writes msbuild Items with their metadata to a props file.
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/src/SaveItems.cs]: >// [](start = 4, length = 2)

Nit: doc comment.

**Comment by safern** [comment]
> The changes look reasonable to me although I'm also curious about what the consumption side in corefx looks like as well.

I'm working on doing that and testing it locally before merging this, I'll open a PR in corefx as well.

**Pull request merged**